zap/src/main/java/org/zaproxy/zap/spider/SpiderController.java

```diff
@@ -251,46 +248,71 @@ public void init() {
 
     /** Clears the previous process. */
     public void reset() {
-        visitedGet.clear();
-        visitedPost.clear();
+        visitedResources.clear();
 
         for (SpiderParser parser : parsers) {
             parser.removeSpiderParserListener(this);
         }
     }
 
-    @Override
-    public void resourceURIFound(
-            HttpMessage responseMessage, int depth, String uri, boolean shouldIgnore) {
-        log.debug("New resource found: " + uri);
+    /**
+     * Builds a canonical identifier for found resources considering the method, URI, headers, and
+     * body.
+     *
+     * @param uri uniform resource identifier for resource
+     * @param resourceFound resource found
+     * @return identifier as a string representation usable for equality checks
+     */
+    private String buildCanonicalResourceIdentifier(URI uri, SpiderResourceFound resourceFound)
+            throws URIException {
+        StringBuilder identifierBuilder = new StringBuilder(50);
+        if (resourceFound != null) {
+            String visitedURI =
+                    URLCanonicalizer.buildCleanedParametersURIRepresentation(
+                            uri,
+                            spider.getSpiderParam().getHandleParameters(),
+                            spider.getSpiderParam().isHandleODataParametersVisited());
+            identifierBuilder.append(resourceFound.getMethod());
+            identifierBuilder.append(" ");
+            identifierBuilder.append(visitedURI);
+            identifierBuilder.append("\n");
+            String canonicalHeaderRepresentation =
+                    getCanonicalHeadersString(resourceFound.getRequestHeaders());
+            identifierBuilder.append("\n");
+            identifierBuilder.append(resourceFound.getBody());
+        }
+        return identifierBuilder.toString();
+    }
 
-        if (uri == null) {
+    @Override
+    public void resourceFound(SpiderResourceFound resourceFound) {
+        if (resourceFound == null
+                || resourceFound.getUri() == null
```

Comment:
Better do these validations when the resource is created, here it should just be used.

Reply:
Resolved via last commit, validations are now done as part of the builder